### ISSUE TITLE
PerlIO_unread fails easily when :crlf is active

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12036]
Migrated from [rt.perl.org#112244](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112244) (status was 'resolved')

Searchable as RT112244$

**Comment by p5pRT** [comment]
### From derhoermi@gmx.net

Hi\,

&nbsp;  I ran into trouble using IO&#8203;::Unread which uses PerlIO\_unread to unread
to a handle\, every now and then it would unread less than what it should
have\. Discussing this on \#p5p it turned out that the :crlf layer was the
problem \(binmode\-ing the handle solved the issue\) and leont asked me to
file a bug about it\. I can supply a test case \(relying on IO&#8203;::Unread\) if
necessary\.

regards\,
\-\- 
Björn Höhrmann · mailto&#8203;:bjoern@&#8203;hoehrmann\.de · http&#8203;://bjoern\.hoehrmann\.de
Am Badedeich 7 · Telefon&#8203;: \+49\(0\)160/4415681 · http&#8203;://www\.bjoernsworld\.de
25899 Dagebüll · PGP Pub\. KeyID&#8203;: 0xA4357E78 · http&#8203;://www\.websitedev\.de/

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Apr 04 10&#8203;:47&#8203;:08 2012\, derhoermi@&#8203;gmx\.net wrote&#8203;:
> Hi\,
> 
>   I ran into trouble using IO&#8203;::Unread which uses PerlIO\_unread to unread
> to a handle\, every now and then it would unread less than what it should
> have\. Discussing this on \#p5p it turned out that the :crlf layer was the
> problem \(binmode\-ing the handle solved the issue\) and leont asked me to
> file a bug about it\. I can supply a test case \(relying on IO&#8203;::Unread\) if
> necessary\.

Please do\.  Thank you\.

A test case using only core modules is ideal\, but one that does not meet
the ideal is better than none\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Leont

On Wed Apr 04 10&#8203;:47&#8203;:08 2012\, derhoermi@&#8203;gmx\.net wrote&#8203;:
> Hi\,
> 
>   I ran into trouble using IO&#8203;::Unread which uses PerlIO\_unread to unread
> to a handle\, every now and then it would unread less than what it should
> have\. Discussing this on \#p5p it turned out that the :crlf layer was the
> problem \(binmode\-ing the handle solved the issue\) and leont asked me to
> file a bug about it\. I can supply a test case \(relying on IO&#8203;::Unread\) if
> necessary\.
> 
> regards\,

The problem is that :crlf doesn't fall back on :pending the way :perlio
does when the unread data doesn't fit into its own buffer\. Instead it
just rejects them\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

On Thu\, Apr 5\, 2012 at 7&#8203;:44 AM\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Please do\.  Thank you\.
>
> A test case using only core modules is ideal\, but one that does not meet
> the ideal is better than none\. :\-\)

I think it should be possible to write a core\-only testcase\, by
repeatedly ungetc\(\)ing a character\. It isn't pretty\, but it should
work\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

On Wed Apr 04 23&#8203;:23&#8203;:51 2012\, LeonT wrote&#8203;:
> On Wed Apr 04 10&#8203;:47&#8203;:08 2012\, derhoermi@&#8203;gmx\.net wrote&#8203;:
> > Hi\,
> > 
> >   I ran into trouble using IO&#8203;::Unread which uses PerlIO\_unread to unread
> > to a handle\, every now and then it would unread less than what it should
> > have\. Discussing this on \#p5p it turned out that the :crlf layer was the
> > problem \(binmode\-ing the handle solved the issue\) and leont asked me to
> > file a bug about it\. I can supply a test case \(relying on IO&#8203;::Unread\) if
> > necessary\.
> > 
> > regards\,
> 
> The problem is that :crlf doesn't fall back on :pending the way :perlio
> does when the unread data doesn't fit into its own buffer\. Instead it
> just rejects them\.
> 
> Leon

I think this should fix it\, but I don't have a proper test just yet\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>0001-Hopefully-fix-112244.patch</summary>

```diff
From 897d1b45865a92d8bb55591a7b5661494b90894c Mon Sep 17 00:00:00 2001
From: Leon Timmermans <fawaka@gmail.com>
Date: Thu, 5 Jul 2012 16:17:44 +0200
Subject: [PATCH] Hopefully fix #112244

:crlf currently doesn't fall back on :pending the way :perlio
does when the unread data doesn't fit into its own buffer. Instead it
just rejects them. This patch resolves that
---
 perlio.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/perlio.c b/perlio.c
index a593c48..a69b1ae 100644
--- a/perlio.c
+++ b/perlio.c
@@ -4586,6 +4586,8 @@ PerlIOCrlf_unread(pTHX_ PerlIO *f, const void *vbuf, Size_t count)
 		}
 	    }
 	}
+        if (count > 0)
+            unread += PerlIOBase_unread(aTHX_ f, vbuf + unread, count);
 	return unread;
     }
 }
-- 
1.7.6.1


```
</details>

**Comment by p5pRT** [comment]
### From @Leont

On Fri\, Jul 6\, 2012 at 11&#8203;:18 AM\, Leon Timmermans via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Wed Apr 04 23&#8203;:23&#8203;:51 2012\, LeonT wrote&#8203;:
>> The problem is that :crlf doesn't fall back on :pending the way :perlio
>> does when the unread data doesn't fit into its own buffer\. Instead it
>> just rejects them\.
>>
>> Leon
>
> I think this should fix it\, but I don't have a proper test just yet\.
>
> Leon

Can someone with more time at hand write some tests for this?
IO&#8203;::Handle's ungetc method looks like the most obvious way to do this\.
I'd like to get this in before 5\.18 but I'm not sure I'll have the
time this month\.

Leon

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by ec1da995fb927cb6d590ede4756eae4e266d82b6

I added some tests that the author thinks are sufficient for the time
being\, in commit f10c05c1d5f850f4bfb45f5d10409f4b2f3cfb2f
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
